File: src/main/scala/com/lucidworks/spark/example/ml/SVMExample.scala
package com.lucidworks.spark.example.ml

import com.lucidworks.spark.SparkApp
import com.lucidworks.spark.analysis.LuceneTextAnalyzer
import org.apache.commons.cli.{CommandLine, Option}
import org.apache.spark.mllib.classification.{SVMWithSGD, SVMModel}
import org.apache.spark.mllib.evaluation.BinaryClassificationMetrics
import org.apache.spark.sql.{SaveMode, SQLContext}
import org.apache.spark.sql.types.{StringType, StructField, StructType}
import org.apache.spark.storage.StorageLevel
import org.apache.spark.{SparkContext, SparkConf}
import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.sql.Row
import org.apache.spark.api.java.function.Function
import org.apache.spark.mllib.feature.HashingTF
import org.apache.spark.mllib.feature.Normalizer
import org.apache.spark.mllib.feature.StandardScaler
import org.apache.spark.mllib.feature.StandardScalerModel
import scala.collection.JavaConverters._
import scala.collection
import com.lucidworks.spark.fusion.FusionMLModelSupport

import scala.collection.immutable

/**
 * A SVM example for text classification. This class has the following feature transformations
 * LuceneAnalyzer
 * HashingTF
 * Normalizer
 * StandardScaler
 * The SVM Model is trained after the above transformations and is tested for the given test data (using same transformations).
 */

object SVMExample {
  val DEFAULT_NUM_FEATURES = "1000000"
  val DEFAULT_NUM_ITERATIONS = "200"
  val DefaultZkHost = "localhost:9983"
  val DefaultCollection = "twitter_sentiment"
}

class SVMExample extends SparkApp.RDDProcessor  {

  import SVMExample._
  def getName = "mllib-svm-scala"
  def getOptions = Array(
    Option.builder().longOpt("indexTrainingData").hasArg.required(false).desc(
      s"Path to training data to index").build(),
    Option.builder().longOpt("indexTestData").hasArg.required(false).desc(
      s"Path to test data to index").build(),
    Option.builder().longOpt("sample").hasArg.required(false).desc(
      s"Fraction (0 to 1) of full dataset to sample from Solr, default is 1").build(),
    Option.builder().longOpt("numFeatures").hasArg.required(false).desc(
      s"Number of features; default is $DEFAULT_NUM_FEATURES").build(),
    Option.builder().longOpt("numIterations").hasArg.required(false).desc(
      s"Number of iterations; default is $DEFAULT_NUM_ITERATIONS").build(),
    Option.builder().longOpt("modelOutput").hasArg.required(false).desc(
      s"Model output path; default is mllib-svm-sentiment").build(),
    Option.builder().longOpt("fusionHostAndPort").hasArg.required(false).desc(
      s"Fusion host and port; Example localhost:8764").build(),
    Option.builder().longOpt("fusionUser").hasArg.required(false).desc(
      s"Fusion user name").build(),
    Option.builder().longOpt("fusionPassword").hasArg.required(false).desc(
      s"Fusion password").build(),
    Option.builder().longOpt("fusionRealm").hasArg.required(false).desc(
      s"Fusion Realm").build()
  )

  override def run(conf: SparkConf, cli: CommandLine): Int = {
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc)
    val csvSchema = StructType(StructField("polarity",StringType, true) ::
      StructField("id",StringType, true) ::
      StructField("date",StringType, true) ::
      StructField("query",StringType, true) ::
      StructField("username",StringType, true) ::
      StructField("tweet_txt",StringType, true) :: Nil)

    val writeoptions = immutable.HashMap(
      "zkhost" -> cli.getOptionValue("zkHost", DefaultZkHost),
      "collection" -> cli.getOptionValue("collection", DefaultCollection),
      "soft_commit_secs" -> "10")

    val indexTrainingData = cli.getOptionValue("indexTrainingData")
    if (indexTrainingData != null) {
      var csvDF = sqlContext.read.format("com.databricks.spark.csv").schema(csvSchema).option("header", "false").load(indexTrainingData)
      csvDF = csvDF.repartition(4)

      csvDF.write.format("solr").options(writeoptions).mode(SaveMode.Overwrite).save()
    }

    val indexTestData = cli.getOptionValue("indexTestData");
    if (indexTestData != null) {
      var csvDF = sqlContext.read.format("com.databricks.spark.csv").schema(csvSchema).option("header", "false").load(indexTestData)
      csvDF = csvDF.withColumnRenamed("polarity", "test_polarity")

      csvDF.write.format("solr").options(writeoptions).mode(SaveMode.Overwrite).save()
    }

    val contentFields = "tweet_txt"

    var trainoptions = immutable.HashMap(
      "zkhost" -> cli.getOptionValue("zkHost", DefaultZkHost),
      "collection" -> cli.getOptionValue("collection", DefaultCollection),
      "query" -> "+polarity:(0 OR 4) +tweet_txt:[* TO *]",
      "fields" -> "id,polarity,tweet_txt",
      "rows" -> "10000",
      "splits" -> "true",
      "split_field" -> "_version_",
      "splits_per_shard" -> "8")

    val sampleFraction = cli.getOptionValue("sample", "1.0").toDouble
    var trainingDataFromSolr = sqlContext.read.format("solr").options(trainoptions).load()
    trainingDataFromSolr = trainingDataFromSolr.sample(false, sampleFraction)

    val inputCols = contentFields.split(" ").map(_.trim)

    val stdTokLowerSchema = "{ \"analyzers\": [{ \"name\": \"std_tok_lower\", \"tokenizer\": { \"type\": \"standard\" },\n" +
      "                \"filters\": [{ \"type\": \"lowercase\" }]}],\n" +
      "  \"fields\": [{ \"regex\": \".+\", \"analyzer\": \"std_tok_lower\" }]}\n"

    val numFeatures = cli.getOptionValue("numFeatures", DEFAULT_NUM_FEATURES).toInt
    val numIterations = cli.getOptionValue("numIterations", DEFAULT_NUM_ITERATIONS).toInt

    def RowtoLab(row: Row, numFeatures: Int, inputCols: Array[String], stdTokLowerSchema: String ): LabeledPoint = {
      var textAnalyzer: LuceneTextAnalyzer = new LuceneTextAnalyzer(stdTokLowerSchema)
      var hashingTF = new HashingTF(numFeatures)
      var normalizer = new Normalizer()
      val polarity = row.getString(row.fieldIndex("polarity"))
      var fields = new java.util.HashMap[String, String]()
      for(i <- 0 until inputCols.length){
        val value = row.getString(row.fieldIndex(inputCols(i)))
        if (value != null) {
          fields.put(inputCols(i), value)
        }
      }
      val analyzedFields = textAnalyzer.analyzeJava(fields)
      var terms = new java.util.LinkedList[String]()
      analyzedFields.values().asScala.toList.foreach(v => terms.addAll(v))
      val sentimentLabel = if (("0" == polarity)) 0.toDouble else 1.toDouble
      new LabeledPoint(sentimentLabel, normalizer.transform(hashingTF.transform(terms)))
    }

    var trainingData = trainingDataFromSolr.rdd.map(row => RowtoLab(row, numFeatures, inputCols, stdTokLowerSchema))
    var standardScaler = new StandardScaler().fit(trainingData.map(x => x.features))
    var trainRDD = trainingData.map(x => new LabeledPoint(x.label, standardScaler.transform(x.features)))
    trainRDD = trainRDD.persist(StorageLevel.MEMORY_ONLY_SER)
    val model = SVMWithSGD.train(trainRDD, numIterations)

    var testoptions = immutable.HashMap("zkhost" -> cli.getOptionValue("zkHost", DefaultZkHost),
      "collection" -> cli.getOptionValue("collection", DefaultCollection),
      "query" -> "+test_polarity:[* TO *] +tweet_txt:[* TO *]",
      "fields" -> "id,test_polarity,tweet_txt")

    var testDataFromSolr = sqlContext.read.format("solr").options(testoptions).load()
    testDataFromSolr = testDataFromSolr.withColumnRenamed("test_polarity", "polarity")
    testDataFromSolr.show
    val testVectors = testDataFromSolr.rdd.map(row => RowtoLab(row, numFeatures, inputCols, stdTokLowerSchema)).map(x => new LabeledPoint(x.label, standardScaler.transform(x.features)))

    val scoreAndLabels =  testVectors.map(p => {
      val score = model.predict(p.features)
      println(">> model predicted: " + score + ", actual: " + p.label)
      new (Double, Double)(score, p.label)
    })

    val metrics = new BinaryClassificationMetrics(scoreAndLabels)
    val auROC = metrics.areaUnderROC
    println("Area under ROC = " + auROC)

    if (cli.getOptionValue("fusionHostAndPort") != null) {
      var metadata = new java.util.HashMap[String, String]()
      metadata.put("numFeatures", "1000000")
      metadata.put("featureFields", "tweet_txt")
      metadata.put("analyzerJson", stdTokLowerSchema)
      metadata.put("normalizer", "Y")
      metadata.put("standardscaler", "Y")
      metadata.put("mean", standardScaler.mean.toString)
      metadata.put("std", standardScaler.std.toString)

      if (cli.getOptionValue("fusionUser") != null && cli.getOptionValue("fusionPassword") != null) {
        FusionMLModelSupport.saveModelInFusion(cli.getOptionValue("fusionHostAndPort"), cli.getOptionValue("fusionUser"), cli.getOptionValue("fusionPassword"),
          cli.getOptionValue("fusionRealm", "native"), sc, cli.getOptionValue("modelOutput", "mllib-svm-sentiment"), model, metadata)
      } else {
        FusionMLModelSupport.saveModelInLocalFusion(sc, cli.getOptionValue("modelOutput", "mllib-svm-sentiment"), model, metadata)
      }
    }
    else {
      model.save(sc, cli.getOptionValue("modelOutput", "mllib-svm-sentiment"))
    }

    return 0
  }

}

File: src/main/java/com/lucidworks/spark/example/ml/SVMExample.java
package com.lucidworks.spark.example.ml;

import com.lucidworks.spark.SparkApp;
import com.lucidworks.spark.analysis.LuceneTextAnalyzer;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Option;
import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.mllib.classification.SVMModel;
import org.apache.spark.mllib.classification.SVMWithSGD;
import org.apache.spark.mllib.evaluation.BinaryClassificationMetrics;
import org.apache.spark.mllib.feature.HashingTF;
import org.apache.spark.mllib.linalg.*;
import org.apache.spark.mllib.linalg.Vector;
import org.apache.spark.mllib.regression.LabeledPoint;
import org.apache.spark.rdd.RDD;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.SaveMode;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.Metadata;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import org.apache.spark.storage.StorageLevel;
import scala.Tuple2;
import org.apache.spark.mllib.feature.Normalizer;
import org.apache.spark.mllib.feature.StandardScaler;
import org.apache.spark.mllib.feature.StandardScalerModel;

import java.io.Serializable;
import java.util.*;

public class SVMExample implements SparkApp.RDDProcessor {

  private static final String DEFAULT_NUM_FEATURES = "1000000";
  private static final String DEFAULT_NUM_ITERATIONS = "200";

  class RowToLabeledPoint implements Serializable, Function<Row, LabeledPoint> {

    private String[] inputCols;
    private String textAnalyzerJson;
    private HashingTF hashingTF;
    private Normalizer normalizer =  new Normalizer();
    private transient LuceneTextAnalyzer textAnalyzer;

    RowToLabeledPoint(String[] inputCols, int numFeatures, String textAnalyzerJson) {
      this.inputCols = inputCols;
      this.hashingTF = new HashingTF(numFeatures);
      this.textAnalyzerJson = textAnalyzerJson;
    }

    public LabeledPoint call(Row row) throws Exception {
      String polarity = row.getString(row.fieldIndex("polarity"));

      Map<String,String> fields = new HashMap<String, String>();
      for (String col : inputCols) {
        String val = row.getString(row.fieldIndex(col));
        if (val != null) {
          fields.put(col, val);
        }
      }

      // eval it lazily to maintain serializability of this Function
      if (textAnalyzer == null) {
        textAnalyzer = new LuceneTextAnalyzer(textAnalyzerJson);
      }

      Map<String,List<String>> analyzedFields = textAnalyzer.analyzeJava(fields);
      List<String> terms = new LinkedList<>();
      for (List<String> termsPerField : analyzedFields.values()) {
        terms.addAll(termsPerField);
      }

      double sentimentLabel = "0".equals(polarity) ? 0d : 1d;
      return new LabeledPoint(sentimentLabel, normalizer.transform(hashingTF.transform(terms)));
    }
  };

  public String getName() {
    return "mllib-svm";
  }

  public Option[] getOptions() {
    return new Option[] {
            Option.builder()
                    .hasArg()
                    .required(false)
                    .desc("Path to training data to index")
                    .longOpt("indexTrainingData")
                    .build(),
            Option.builder()
                    .hasArg()
                    .required(false)
                    .desc("Path to test data to index")
                    .longOpt("indexTestData")
                    .build(),
            Option.builder()
                    .hasArg()
                    .required(false)
                    .desc("Fraction (0 to 1) of full dataset to sample from Solr, default is 1")
                    .longOpt("sample")
                    .build(),
            Option.builder()
                    .hasArg()
                    .required(false)
                    .desc("Number of features; default is " + DEFAULT_NUM_FEATURES)
                    .longOpt("numFeatures")
                    .build(),
            Option.builder()
                    .hasArg()
                    .required(false)
                    .desc("Number of iterations; default is "+DEFAULT_NUM_ITERATIONS)
                    .longOpt("numIterations")
                    .build(),
            Option.builder()
                    .hasArg()
                    .required(false)
                    .desc("Model output path; default is mllib-svm-sentiment")
                    .longOpt("modelOutput")
                    .build()
    };
  }

  public int run(SparkConf conf, CommandLine cli) throws Exception {
    JavaSparkContext jsc = new JavaSparkContext(conf);
    SQLContext sqlContext = new SQLContext(jsc);

    String zkHost = cli.getOptionValue("zkHost", "localhost:9983");
    String collection = cli.getOptionValue("collection", "twitter_sentiment");

    StructType csvSchema = new StructType(new StructField[] {
            new StructField("polarity", DataTypes.StringType, true, Metadata.empty()),
            new StructField("id", DataTypes.StringType, true, Metadata.empty()),
            new StructField("date", DataTypes.StringType, true, Metadata.empty()),
            new StructField("query", DataTypes.StringType, true, Metadata.empty()),
            new StructField("username", DataTypes.StringType, true, Metadata.empty()),
            new StructField("tweet_txt", DataTypes.StringType, true, Metadata.empty())
    });

    // note: get training data from:
    // http://help.sentiment140.com/for-students
    // extract the zip locally and point to it with command-line opts
    //
    // this is just for example purposes as the Area Under ROC from this model is 0.65, which is only slightly
    // better than random :-(
    String indexTrainingData = cli.getOptionValue("indexTrainingData");
    if (indexTrainingData != null) {

      DataFrame csvDF = sqlContext.read()
              .format("com.databricks.spark.csv")
              .schema(csvSchema)
              .option("header", "false")
              .load(indexTrainingData);
      csvDF = csvDF.repartition(4);
      Map<String, String> options = new HashMap<>();
      options.put("zkhost", zkHost);
      options.put("collection", collection);
      options.put("soft_commit_secs", "10");
      csvDF.write().format("solr").options(options).mode(SaveMode.Overwrite).save();
    }

    String indexTestData = cli.getOptionValue("indexTestData");
    if (indexTestData != null) {
      DataFrame csvDF = sqlContext.read()
              .format("com.databricks.spark.csv")
              .schema(csvSchema)
              .option("header", "false")
              .load(indexTestData);
      csvDF = csvDF.withColumnRenamed("polarity","test_polarity");
      Map<String, String> options = new HashMap<>();
      options.put("zkhost", zkHost);
      options.put("collection", collection);
      options.put("soft_commit_secs", "10");
      csvDF.write().format("solr").options(options).mode(SaveMode.Overwrite).save();
    }

    final String contentFields = "tweet_txt";

    Map<String, String> options = new HashMap<>();
    options.put("zkhost", zkHost);
    options.put("collection", collection);
    options.put("query", "+polarity:(0 OR 4) +tweet_txt:[* TO *]");
    options.put("fields", "id,polarity,tweet_txt");
    options.put("rows", "10000");
    options.put("splits", "true");
    options.put("split_field", "_version_");
    options.put("splits_per_shard", "8");

    double sampleFraction = Double.parseDouble(cli.getOptionValue("sample", "1.0"));
    DataFrame trainingDataFromSolr = sqlContext.read().format("solr").options(options).load();
    trainingDataFromSolr = trainingDataFromSolr.sample(false, sampleFraction);
    //trainingDataFromSolr.show();

    final String[] inputCols = contentFields.split(",");
    for (int i = 0 ; i < inputCols.length ; ++i) {
      inputCols[i] = inputCols[i].trim();
    }

    // use the standard tokenizer
    String stdTokLowerSchema = json(
            "{ 'analyzers': [{ 'name': 'std_tok_lower', 'tokenizer': { 'type': 'standard' },\n" +
                    "                'filters': [{ 'type': 'lowercase' }]}],\n" +
                    "  'fields': [{ 'regex': '.+', 'analyzer': 'std_tok_lower' }]}\n");

    final int numFeatures = Integer.parseInt(cli.getOptionValue("numFeatures", DEFAULT_NUM_FEATURES));
    final Function<Row, LabeledPoint> mapFunc = new RowToLabeledPoint(inputCols, numFeatures, stdTokLowerSchema);
    final int numIterations = Integer.parseInt(cli.getOptionValue("numIterations", DEFAULT_NUM_ITERATIONS));
    JavaRDD<LabeledPoint> trainingData = trainingDataFromSolr.javaRDD().map(mapFunc);
    final StandardScalerModel standardScaler = new StandardScaler().fit(trainingData.map(new Function<LabeledPoint, org.apache.spark.mllib.linalg.Vector>() {
      public Vector call(LabeledPoint labpt) {
        return labpt.features();
      }
    }).rdd());
    RDD<LabeledPoint> trainRDD = trainingData.map(new Function<LabeledPoint, LabeledPoint>() {
      public LabeledPoint call(LabeledPoint lp) {
        return new LabeledPoint(lp.label(), standardScaler.transform(lp.features()));
      }
    }).rdd();

    trainRDD = trainRDD.persist(StorageLevel.MEMORY_ONLY_SER());
    final SVMModel model = SVMWithSGD.train(trainRDD, numIterations);

    // Clear the default threshold.
    //model.clearThreshold();

    options = new HashMap<>();
    options.put("zkhost", zkHost);
    options.put("collection", collection);

    // find docs that don't have polarity
    options.put("query", "+test_polarity:[* TO *] +tweet_txt:[* TO *]");
    options.put("fields", "id,test_polarity,tweet_txt");
    DataFrame testDataFromSolr = sqlContext.read().format("solr").options(options).load();
    testDataFromSolr = testDataFromSolr.withColumnRenamed("test_polarity", "polarity");
    testDataFromSolr.show();

    JavaRDD<LabeledPoint> testVectors = testDataFromSolr.javaRDD().map(mapFunc).map(new Function<LabeledPoint, LabeledPoint>() {
      public LabeledPoint call(LabeledPoint lp) {
        return new LabeledPoint(lp.label(), standardScaler.transform(lp.features()));
      }
    });

    // Compute raw scores on the test set.
    JavaRDD<Tuple2<Object, Object>> scoreAndLabels = testVectors.map(
      new Function<LabeledPoint, Tuple2<Object, Object>>() {
        public Tuple2<Object, Object> call(LabeledPoint p) {
          Double score = model.predict(p.features());
          System.out.println(">> model predicted: "+score+", actual: "+p.label());
          return new Tuple2<Object, Object>(score, p.label());
        }
      }
    );

    // Get evaluation metrics.
    BinaryClassificationMetrics metrics = new BinaryClassificationMetrics(JavaRDD.toRDD(scoreAndLabels));
    double auROC = metrics.areaUnderROC();
    System.out.println("Area under ROC = " + auROC);

    // Save and load model
    model.save(jsc.sc(), cli.getOptionValue("modelOutput", "mllib-svm-sentiment"));
    return 0;
  }

  private String json(String singleQuoted) {
    return singleQuoted.replaceAll("'", "\"");
  }

}
